File: C++/maximum-frequency-stack.cpp
// Time:  O(1)
// Space: O(n)

class FreqStack {
public:
    FreqStack() 
        : max_freq_(0) {
        
    }
    
    void push(int x) {
        ++freq_[x];
        if (freq_[x] > max_freq_) {
            max_freq_ = freq_[x];
        }
        group_[freq_[x]].emplace_back(x);
    }
    
    int pop() {
        auto x = group_[max_freq_].back(); group_[max_freq_].pop_back();
        if (group_[max_freq_].empty()) {
            group_.erase(max_freq_);
            --max_freq_;
        }
        --freq_[x];
        if (freq_[x] == 0) {
            freq_.erase(x);
        }
        return x;
    }

private:
    unordered_map<int, int> freq_;
    unordered_map<int, vector<int>> group_;
    int max_freq_;
    
};

/**
 * Your FreqStack object will be instantiated and called as such:
 * FreqStack obj = new FreqStack();
 * obj.push(x);
 * int param_2 = obj.pop();
 */

File: Python/maximum-frequency-stack.py
# Time:  O(1)
# Space: O(n)

import collections


class FreqStack(object):

    def __init__(self):
        self.__freq = collections.Counter()
        self.__group = collections.defaultdict(list)
        self.__maxfreq = 0

    def push(self, x):
        """
        :type x: int
        :rtype: void
        """
        self.__freq[x] += 1
        if self.__freq[x] > self.__maxfreq:
            self.__maxfreq = self.__freq[x]
        self.__group[self.__freq[x]].append(x)   

    def pop(self):
        """
        :rtype: int
        """
        x = self.__group[self.__maxfreq].pop()
        if not self.__group[self.__maxfreq]:
            self.__group.pop(self.__maxfreq)
            self.__maxfreq -= 1
        self.__freq[x] -= 1
        if not self.__freq[x]:
            self.__freq.pop(x)
        return x



